Deadline Integration with Enhanced Features and Bug Fixes

RyanLloydZA opened the pull request:
Introduced Deadline Submission and UI Interaction with Progress Bar and Item Status Updates.

Right Click Menu clean up
Multiple small bug fixs
Some Code Clean up and refactoring

It Still needs some more additional clean up and some variable renaming but i think its in an ok state until the next features or possible bugs get addressed

RyanLloydZA committed:
Introduced Deadline Submission and UI Interaction with Progress Bar and Item Status Updates.

Right Click Menu cleanup
Multiple small bug fixs
Some Code Clean up and refactoring

It Still needs some more additional cleanup and some variable renaming but i think its in an ok state until the next features or possible bugs get addressed

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -71,17 +79,62 @@ def __init__(self):
+        # Set the window size
+        self.resize(1000, self.ui.compTableWidget.sizeHint().height())
+
+    def get_sg_fields(self):
```
the "get" in this function name kindve implies that it's returning a value. Using `populate` like you did for the deadline settings probably makes more sense.

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -71,17 +79,62 @@ def __init__(self):
         # lastly, set up our very basic UI
         # self.ui.context.setText("Current Context: %s" % self._app.context)
         self.populate_presets()
+        self.populate_deadline_settings()
```
`DeadlineCommand` which this is using can be quite slow. Even if it's quick now, as soon as we have any network slowness or the deadline machine is bogged down, it will get INCREDIBLY slow. 

Since this UI has multiple purposes beyond just submitting jobs to deadline, I think we should try to lazily initialize these values. 
This way, the UI is only affected by the DeadlineCommand slowness when the artist actually intends to use Deadline

JacobTerritory replied:
I don't remember the exact layout of your UI, but I think the deadline part is expandable/collapsible, so we can probably initialize all this stuff when the UI is expanded.

RyanLloydZA replied:
Currently the UI Opens with it expanded already so i could change that to open collapsed, would also need to check and initialize it on submission to if its never opened.

JacobTerritory replied:
Yea, that is fine - The point is to delay the slow DeadlineCommand calls for as long as possible

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -351,6 +432,9 @@ def apply_to_render_queue_items(self):
             includeCheckBox = self.ui.compTableWidget.item(row, 6)
 
+            # Emit progress
+            self.update_progress_bar(int((row + 1) / self.ui.compTableWidget.rowCount() * 100))
```
I noticed this in the UI when you were demoing it... It felt like the progress bar was 1 element ahead of where it should be.
This update should probably be at the end of this block of code (Or remove the +1)

RyanLloydZA replied:
Removing the +1 Could help that, Currently every row has to get processed even if it doesnt pas the checks, so moving this down would need duplication of the progress change where it does skip to the next item or it would look like its skipping ahead/lagging behind when items are skipped until it gets to one that passes again.

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+        logger.debug("Project Checks passed, proceeding with submission")
+
+        # Save and create a copy of the current project
+        self.adobe.app.project.save()
```
Unprompted saving can be a little annoying. Perhaps a prompt first? (Speaking from past experience with artists)

Also, aren't you already checking if the project is saved up above?

Ahhh the above code is checking if the project has been saved to a path yet (As in, is it new)... Is there a way of knowing whether or not the project file currently has unsaved edits? If that's not trivial then this is probably fine as it is

JacobTerritory replied:
Also is saving a project in AE ever slow? What happens if you save a project which doesn't need saving? Does it still go through all the motions and take the extra time anyway?

RyanLloydZA replied:
Yeah, there is no way using jsx script to check the current state of the file. Can  only check if its been saved to disk, Sometimes depending on the size of the project it can be slow. This is something we can discuss with them, because it comes down to using this as a fail safe, introducing a prompt which would resolve that but could cause annoyance or leave it on the artists to make sure what they are sending is up to date and saved.

JacobTerritory replied:
Yea, a prompt to save when you've already saved is probably more annoying than just forcing a save. 

Lets leave it as is and see what people say.

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+
+        except Exception as e:
+            logger.error("Failed to create backup: %s" % e)
+            traceback.print_exc()
```
Everywhere else you are using the logger - Might be good to use the logger here as well.

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+        for row in range(self.ui.compTableWidget.rowCount()):
+
+            # Emit progress
+            self.update_progress_bar(int((row + 1) / self.ui.compTableWidget.rowCount() * 100))
```
Again with the (imo) pre-mature progress bar updates

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+        return deadlineCommand
+
+    def call_deadline_command(self, arguments, hideWindow=True, readStdout=True):
+        """
+        Calls the Deadline command with specified arguments.
+
+        Args:
+            arguments (list): List of arguments to pass to the Deadline command.
+            hideWindow (bool): Whether to hide the command window (Windows only).
+            readStdout (bool): Whether to read the standard output.
+
+        Returns:
+            str: The standard output from the command if readStdout is True.
```
This is a different doc string format as all your others (I assume it's because you just copy pastad this - Though I do prefer this method - Feels more human readable)

RyanLloydZA replied:
Yeah this was added formt he Houdini call deadline command which i changed the logic of anyways. Yeah i dont normally  add the type to the arguments in brackets as its normally mentioned in the description regardless. But i get the appeal for fast browsing over the code

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+
+        # Check if the output module is rendering a movie
+        is_movie =  False
+        movie_formats = self.deadline_defaults["movie_formats"]
```
This code should probably be:
```
_, ext = os.path.splitext(output_file_name)
is_movie = ext.lower() in movie_formats
```

JacobTerritory replied:
(Though I'm not sure how this handles paths with multiple dots in the extension...but that isn't really relevant here)

RyanLloydZA replied:
came up with a better execution of my original with not having to worry about splitting it correctly, 
`is_movie = any(output_file_name.endswith(ext) for ext in self.deadline_defaults["movie_formats"])`

Had forgotten about the any function

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+            frame_offset = self.adobe.app.project.displayStartFrame
+            display_start_time = render_queue_item.comp.displayStartTime
+
+            if display_start_time:
+                start_frame = frame_offset + round(display_start_time / frame_duration) + round(render_queue_item.timeSpanStart / frame_duration)
+                end_frame = start_frame + round(render_queue_item.timeSpanDuration / frame_duration) - 1
+                frame_list = f"{start_frame}-{end_frame}"
```
This logic looks really weird to me. 

Is `frame_duration` a floating point variable to tweak how "long" a single frame is? So it's basically a way of applying a flat retime curve to everything?

RyanLloydZA replied:
Yeah it is a float value as after effects doesnt work with frame values on its output, So the start and end frame values have to be worked out from the duration of the time spans selected and the start time of the comps. THis is the direct approach used in the original deadline script. My logic doing this in the setup of the time spans for custom in and outs for the templates handle it slightly differently but it gives the same results. I Kept this constant to the original file

JacobTerritory replied:
Ohhhh, is `frame_duration` something like 24.0 then for 24 fps, and then display_start_time, is the time in seconds?

A comment here explaining the intention + example would be really helpful.

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+        multi_machine = deadline_settings.get('multi_machine', False)
+        submit_scene = deadline_settings.get('submit_scene', False)
+        comp_name = render_queue_item.comp.name
+        dependent_job_id = previous_job_id if previous_job_id else ""
```
This works the same as what you have: `previous_job_id or ""`

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+            submit_info.write(f"Name={job_name}\n")
+            if dependent_comps:
+                submit_info.write(f"BatchName={project_name}\n")
+            submit_info.write(f"Comment={deadline_settings.get('comment', '')}\n")
+            submit_info.write(f"Department={deadline_settings.get('department', '')}\n")
+            submit_info.write(f"Group={deadline_settings.get('group', 'ae')}\n")
+            submit_info.write(f"Pool={deadline_settings.get('pool', 'none')}\n")
+            submit_info.write(f"SecondaryPool={deadline_settings.get('secondary_pool', 'none')}\n")
+            submit_info.write(f"Priority={deadline_settings.get('priority', 30)}\n")
+            submit_info.write(f"TaskTimeoutMinutes={deadline_settings.get('task_timeout_minutes', '0')}\n")
+            submit_info.write(f"LimitGroups={deadline_settings.get('limit_groups', '')}\n")
+            submit_info.write(f"ConcurrentTasks={deadline_settings.get('concurrent_tasks', '1')}\n")
+            submit_info.write(f"LimitConcurrentTasksToNumberOfCpus={deadline_settings.get('limit_concurrent_tasks_to_number_cpus', '0')}\n")
+            submit_info.write(f"JobDependencies={current_job_dependencies}\n")
+            submit_info.write(f"OnJobComplete={deadline_settings.get('on_job_complete', 'Nothing')}\n")
```
I don't really think this should be set here, it should more-so be a global setting set in Deadline, but the Job Error Limit should be set to something reasonable like 50 rather than 500... As it stands now an erroring job on Deadline will sit there and error for an entire day before actually failing. 
And it will take up a bunch of nodes from other jobs in the process.

RyanLloydZA replied:
Sure that could be removed but it was included to match the functionality as is. Error limited I Will take a look at

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+
+            # Output file for all output modules
+            index = 0
+            for i in range(1, render_queue_item.numOutputModules + 1):
+                output_module = render_queue_item.outputModule(i)
```
Does AE seriously not count from 0...

RyanLloydZA replied:
It does not, they index from 1

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+                submit_info.write(f"Frames={frame_list}\n")
+
+            # Output file for all output modules
+            index = 0
```
You are using index here, but never iterating it.

RyanLloydZA replied:
Yes because currently we are still only supporting a single output module, since that was partly discussed iv placed it here for use if they decide they would want to move forward with that workflow in the future, a place holder XD

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -921,4 +1039,632 @@ def supress_dialogs(self):
+                submit_info.write(f"MachineLimit=1\n")
+                submit_info.write(f"ChunkSize=1000000\n")
+            else:
+                if multi_machine:
+                    submit_info.write(f"MachineLimit=0\n")
+                    submit_info.write(f"ChunkSize=1\n")
```
What is `multi_machine`, and why does it have a chunk size of 1?

JacobTerritory commented on python/app/ui/dialog.py:
```diff
@@ -19,8 +25,512 @@ def __init__(self, parent=None):
+        # Buttons
+        self.button_layout = QtGui.QVBoxLayout(self)
+
+        self.add_button = QtGui.QPushButton(">", self)
+        self.add_button.setToolTip("Add selected items to the selected list")
+        self.add_button.setMaximumHeight(30)
+
+        self.remove_button = QtGui.QPushButton("<", self)
+        self.remove_button.setToolTip("Remove selected items from the selected list")
+        self.remove_button.setMaximumHeight(30)
+
+        self.move_up_button = QtGui.QPushButton("^", self)
+        self.move_up_button.setToolTip("Move selected machine up")
+        self.move_up_button.setMaximumHeight(30)
+
+        self.move_down_button = QtGui.QPushButton("v", self)
+        self.move_down_button.setToolTip("Move selected machine down")
+        self.move_down_button.setMaximumHeight(30)
```
Would be cool if you also bound these to the corresponding button presses... But that can be a future thing :p

RyanLloydZA replied:
I like that.. nice idea!

JacobTerritory commented on python/app/ui/dialog.py:
```diff
@@ -19,8 +25,512 @@ def __init__(self, parent=None):
+        """
+        self.selected_list.clear()
+        self.selected_list.addItems(selected_items)
+        logger.debug(" selected List Updated: %s", selected_items)
```
Random extra white space here

JacobTerritory commented on python/app/ui/dialog.py:
```diff
@@ -19,8 +25,512 @@ def __init__(self, parent=None):
+            Populate the available and selected lists from a widget.
+        """
+        logger.debug("Populating item lists from widget")
+        if isinstance(self.widget, (LineEditButton, QtGui.QLineEdit)):
+            item_list = self.string_to_item_list(
+                self.widget.get_text() if isinstance(self.widget, LineEditButton) else self.widget.text())
+            self.populate_selected_list(item_list)
```
This logic looks weird, and is hard to read. Something like this might be better
```
if isinstance(self.widget, LineEditButton):
    text = self.widget.get_text()
elif isinstance(self.widget, QtGui.QLineEdit)
    text = self.widget.text()
else: # is an else required here?
    pass
item_list = self.string_to_item_list(text)
self.populate_selected_list(item_list)
```

Alternatively, the LineEditButton is our own code - We could (any maybe should?) update it to follow Qt's conventions. This way we don't need any of this isinstance checking

RyanLloydZA replied:
Yeah this falls under my comment on the initial commit about coming back to rename and clean up some variables, i do agree with this. Initially this was going to be used on the line edit sliders where the values were initially pulled from a slider value and not the text field. But that's irrelevant now

JacobTerritory commented on python/app/ui/dialog.py:
```diff
@@ -19,8 +25,512 @@ def __init__(self, parent=None):
+        """
+        selected_string = self.selected_items_to_string()
+
+        if isinstance(self.widget, LineEditButton):
+            self.widget.set_text(selected_string)
+        elif isinstance(self.widget, QtGui.QLineEdit):
+            self.widget.setText(selected_string)
+        else:
+            pass
+        self.close()
```
Ahh! I just had to read 5 more lines! You did it better here lol

Same comment again about maybe matching the Qt Conventions for ease of use. Though it feels dirtier here because they are using camelCase... so maybe not 🤷‍♂️

RyanLloydZA committed:
Enhanced UI functionality:
 Changed the deadline settings panel to a collapsed state on open.
 Multi-selection for available items, and shortcuts for item manipulation in the machine limit and group dialogs

 Introduced a comment field for Deadline submissions and improved the population of Deadline settings and added lazy loading for the deadline calls to only happen when the panel is opened or when the submission button is activated.

 Added functionality to disable buttons once one has been pressed until its function is complete to avoid doubling up by accident.

 Added Failure detection amount for task errors

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -1244,14 +1277,25 @@ def submit_to_deadline(self):
-        # Save and create a copy of the current project
-        self.adobe.app.project.save()
-        logger.info('Project saved')
+        # Open dialog to ask the user if they want to save the project
+        save_project = QtGui.QMessageBox.question(
+            self,
+            "Save Project",
+            "Do you want to save the project before submitting to Deadline?",
+            QtGui.QMessageBox.Yes | QtGui.QMessageBox.No,
+            QtGui.QMessageBox.Yes,
+        )
+        if save_project == QtGui.QMessageBox.Yes:
+            # Save and create a copy of the current project
+            self.adobe.app.project.save()
+            logger.info('Project saved')
+        else:
+            logger.info('Project not saved')
```
As we chatted, lets move this to earlier in the process so the user doesn't need to wait before confirming.

(Or remove it entirely - I'm on the fence)

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -1290,7 +1335,7 @@ def submit_to_deadline(self):
 
             # Emit progress
-            self.update_progress_bar(int((row + 1) / self.ui.compTableWidget.rowCount() * 100))
+            self.update_progress_bar(int(row / self.ui.compTableWidget.rowCount() * 100))
```
Now that you've removed the plus one, the progress bar will never reach 100%.

You could add a simple call after the for loop to just set the progress bar to 100 before closing it.

I added this code to the Ingest publisher:

```
        self.progress_bar.setValue(100)

        # Sleep for a moment before resetting the progress bar to give the UI a chance to draw 100% complete. 
        time.sleep(0.05)
        self.progress_bar.reset()
```

JacobTerritory replied:
The sleep might be a bit extra, but it felt weird never seeing the progress bar hit 100%

JacobTerritory commented on python/app/dialog.py:
```diff
@@ -1458,6 +1500,7 @@ def submit_render_queue_item_to_deadline(self, render_queue_item, deadline_setti
             submit_info.write(f"LimitConcurrentTasksToNumberOfCpus={deadline_settings.get('limit_concurrent_tasks_to_number_cpus', '0')}\n")
             submit_info.write(f"JobDependencies={current_job_dependencies}\n")
             submit_info.write(f"OnJobComplete={deadline_settings.get('on_job_complete', 'Nothing')}\n")
+            submit_info.write(f"FailureDetectionTaskErrors={deadline_settings.get('failure_detection_task_errors', 5)}\n")
```
Is this 5 errors on a single frame chunk? 
There is also a `FailureDetectionJobErrors` property which is global for the whole job. Might be worth setting both values.

With how unreliable our AE farm is, this feels a little low.

We quite frequently have cases where a machine which is bad and will sit and error on a job until it's black-listed. So we need to strike a balance between the black-list detection limit, and the failure limit. 

But that said - Maybe lets leave this as is and tweak as we feel is necessary.

JacobTerritory commented on python/app/ui/dialog.py:
```diff
@@ -283,18 +288,22 @@ def __init__(self, parent=None, item_list=None, widget=None):
         self.add_button = QtGui.QPushButton(">", self)
         self.add_button.setToolTip("Add selected items to the selected list")
         self.add_button.setMaximumHeight(30)
+        self.add_button.setShortcut(QtGui.QKeySequence(QtCore.Qt.Key_Right))
```
Oh! Keyboard shortcuts are so easy. Nicely done!

RyanLloydZA committed:
Refactored Save before Deadline submission so the dialog opens at the beginning of the process and updated item list population logic when pulling info from a widget.

Also added a small sleep timer to show the progress bar at 100%
